### PR TITLE
Drop automagical endpoint calculation, replace with explicit endpoints

### PR DESCRIPTION
**Pull request by nwithan8**
# Description

I've never been a fan of how this (and some of our other libraries) determine what endpoint to hit in our API by basically taking the name of the class, figuring out how to make it plural and then lowercasing it. The "class URL" and "instance URL" utility functions are clunky, do unnecessary reflection, are hard to follow, and are simply begging to break one day in the future simply because a class name does not one-to-one match an endpoint on the API.

Instead, as was done in the C# library during its rewrite last year, this "automagical" utility was removed and instead replaced with explicit endpoint strings. This not only returns transparency to what endpoint is being hit on each function, but also dramatically improves readability of our code.

This PR also makes minor tweaks to how the requestor function operates under-the-hood, mainly by containing client API base and version information solely to that class. Previously, the whole URL (including base and API version) were determined in each service. Now, that calculation is pushed downstream into one location inside the actual request function. This restructuring also allows for the introduction of API version override on a per-request basis, making "beta" endpoints more obvious and meaning less changes needed when a beta features migrates to general availability and no longer needs to hit the "beta" API version.

- Remove endpoint determination by class, replace with explicit endpoint strings
- Modify Requestor under-the-hood to handle new endpoint strings
- Add support for overriding API version on a per-request basis using new endpoint strings

# Testing

- All unit tests pass with no re-recording necessary, implying that the change does not introduce a breaking change to the end-user.
  - Modifications were needed to the mocking rules for Billing unit tests to account for the changes made to the request function downstream. This did not affect the actual functionality or assertions of those unit tests.

# Pull Request Type

Please select the option(s) that are relevant to this PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement (fixing a typo, updating readme, renaming a variable name, etc)

### CODE REVIEW

File: src/main/java/com/easypost/http/Requestor.java
Comment (jchen293) [inline src/main/java/com/easypost/http/Requestor.java:469]: I am fine with this PR. The only concern I have is this public function because it takes in the `apiVersion` parameter, meaning that users can use this function with any `apiVersion`. I believe people will probably not use this `request` function anyway but just wanted to bring this up.

File: src/main/java/com/easypost/service/BatchService.java
Comment (nwithan8) [reply src/main/java/com/easypost/service/BatchService.java:95]: Cleaner vs. waste of a variable, I'll defer to you.

File: src/main/java/com/easypost/http/Requestor.java
Comment (nwithan8) [reply src/main/java/com/easypost/http/Requestor.java:469]: Fair enough, I'll take a crack at making this function inaccessible to end-users since, you're right, no one should be manually calling this.

Comment (nwithan8) [reply src/main/java/com/easypost/http/Requestor.java:469]: I don't think this will be a problem, since, as you said, users shouldn't ever have a need to call this method. But if they did, they could technically hit a different endpoint, as well as a different API version. What I'm saying is, the additional API version parameter doesn't introduce any more risk or anti-pattern than was already present.

I would love for this to be all internal-only access, but as we've discussed before, this method has to be public since it is in a different package/namespace than the services. Such is Java.

File: src/main/java/com/easypost/service/BatchService.java
Comment (Justintime50) [reply src/main/java/com/easypost/service/BatchService.java:95]: I won't block on it, but I don't think it's actually cleaner since you're cramming string interpolation into a function call instead of only giving it finalized parameters.

Comment (nwithan8) [reply src/main/java/com/easypost/service/BatchService.java]: Resolved

File: src/main/java/com/easypost/http/Requestor.java
Comment (jchen293) [reply src/main/java/com/easypost/http/Requestor.java:469]: We should avoid this if we could. One way to solve this is by passing another client with `Beta` apiVersion when initializing `BetaReferralCustomerService` in the EasyPostClient class constructor. This way, we can remove this function.

Comment (nwithan8) [reply src/main/java/com/easypost/http/Requestor.java:469]: Yeah, we should do it that way, I believe that was part of the spec for our rewrite design on how to handle beta functionality, but I'm not seeing a `BetaClient` in the library (a la C#: https://github.com/EasyPost/easypost-csharp/blob/master/EasyPost/BetaClient.cs)

Thoughts @Justintime50 ?

Comment (Justintime50) [reply src/main/java/com/easypost/http/Requestor.java:469]: My suggestion would be to take the path of least resistance. In PHP for instance, we simply have a beta flag on the requestor function that tells it to hit the beta endpoint instead. I would recommend going a similar route.

Comment (nwithan8) [reply src/main/java/com/easypost/http/Requestor.java:469]: Good, so what we have here is good to go.

**Review by Justintime50 — APPROVED** [review]
This is a classic "If it ain't broke, don't fix it" example. I'm not convinced we got value out of the these changes - they also don't really hurt to have either. Seems fine to me.

I would caution though that what efforts we choose to pursue and code that gets refactored should be focused on what returns the most value to us and especially end-users. This is fine, though now fragments how our libs handle endpoint params and wasn't explicitly necessary since the previous implementation has worked great for the last ~8 years.

**Review by jchen293 — APPROVED** [review]

**Pull request merged**
